src/solvers/dg_unstructured_quad/dg_2d.jl

```diff
@@ -307,6 +307,12 @@ end
 function calc_boundary_flux!(cache, t, boundary_conditions,
                              mesh::Union{UnstructuredQuadMesh, P4estMesh},
                              equations, dg::DG)
+  if nboundaries(cache.boundaries) != nboundaries(boundary_conditions)
+    # When using AMR, the boundary container may have been resized.
+    # In this case, the boundary types container must be reinitialized as well.
+    initialize!(boundary_conditions, cache)
```

Comment:
The AMR callback doesn't have the semidiscretization that holds the boundary conditions. Would you prefer passing the boundary conditions to the AMR controller?

Reply:
No, but it might justify passing the `semi` to the AMR controller... what do you think @ranocha? Should we just pass the semi, is it fine as it is, or should we leave it as it is but create an issue for it?

Reply:
I don't have a very strong opinion on this. Passing the `semi` is fine for me (we already pass it to the constructor of the AMR controller).

Reply:
Is this fine, or should I do this via dispatch somehow?
```julia
if semi.boundary_conditions isa UnstructuredQuadSortedBoundaryTypes
  # Reinitialize boundary types container because boundaries may have changed.
  initialize!(semi.boundary_conditions, cache)
end
```

Reply:
I prefer dispatch to make it extendable from outside.